### PR TITLE
apiResolverFactory refactor

### PR DESCRIPTION
**Pull request by axenteoctavian**
## Reasoning behind the pull request
- MX-14961 task
- refactor for file `factory/api/apiResolverFactory.go`
- 
  
## Proposed changes
- refactor function createScQueryElement
- 
- 

## Testing procedure
- running sovereign node
- running main node
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: factory/api/apiResolverFactory.go [hunk 1]
```diff
@@ -406,37 +416,57 @@ func createScQueryElement(
...
 	log.Debug("maximum gas per VM Query", "value", maxGasForVmQueries)
 
 	vmContainer, err := vmFactory.Create()
 	if err != nil {
-		return nil, err
+		return smartContract.ArgsNewSCQueryService{}, err
+	}
+
+	if args.chainRunType == common.ChainRunTypeSovereign {
+		err = addSystemVMToContainer(args, vmContainer, argsHook)
```
Comment (mariusmihaic) [inline factory/api/apiResolverFactory.go:453]: Inside this if you need to to actually do `argsHook.Accounts = accAdapter`

**Review by mariusmihaic — CHANGES_REQUESTED** [review]

File: factory/api/apiResolverFactory.go [hunk 2]
```diff
@@ -406,37 +416,57 @@ func createScQueryElement(
 		MissingTrieNodesNotifier: syncer.NewMissingTrieNodesNotifier(),
 	}
 
-	var apiBlockchain data.ChainHandler
+	apiBlockchain, err := blockchain.NewMetaChain(disabled.NewAppStatusHandler())
+	if err != nil {
+		return smartContract.ArgsNewSCQueryService{}, err
+	}
+	argsHook.BlockChain = apiBlockchain
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply factory/api/apiResolverFactory.go:423]: Actually, for our use case, you only need to pass arround `accAdapter` pointer.
apiBlockChain creation should remain specific for each func inside of them

**Review by sasurobert — DISMISSED** [review]

File: factory/api/apiResolverFactory.go [hunk 3]
```diff
@@ -407,25 +417,56 @@ func createScQueryElement(
 	}
 
 	var apiBlockchain data.ChainHandler
+	var accAdapter state.AccountsAdapterAPI
```
Comment (mariusmihaic) [inline factory/api/apiResolverFactory.go:420]: I think these declarations can be removed.
We only need below to assign in `argsHook` the correct pointers

File: factory/api/apiResolverFactory_test.go [hunk 4]
```diff
@@ -452,3 +497,51 @@ func TestCreateApiResolver_createScQueryElement(t *testing.T) {
...
+		require.Nil(t, err)
+		require.NotNil(t, svm)
+		require.Equal(t, "*process.systemVM", fmt.Sprintf("%T", svm))
+
+		wasmvm, err := argsScQueryService.VmContainer.Get(vmFactory.WasmVirtualMachine)
+		require.Nil(t, err)
+		require.NotNil(t, wasmvm)
+		require.Equal(t, "*hostCore.vmHost", fmt.Sprintf("%T", wasmvm))
+	})
+	t.Run("regular chain should NOT add systemVM", func(t *testing.T) {
```
Comment (mariusmihaic) [inline factory/api/apiResolverFactory_test.go:526]: This is a good test for regular shard chain. We should also check the factory for a metachain shard chain in an unit test, if it doesn't already exist

File: factory/api/apiResolverFactory_test.go [hunk 5]
```diff
@@ -526,7 +528,7 @@ func TestCreateApiResolver_createArgsSCQueryService(t *testing.T) {
 	t.Run("regular chain should NOT add systemVM", func(t *testing.T) {
```
Comment (mariusmihaic) [inline factory/api/apiResolverFactory_test.go:528]:
```suggestion
	t.Run("regular chain for shards should only add wasm vm", func(t *testing.T) {
```

File: factory/api/apiResolverFactory_test.go [hunk 6]
```diff
@@ -544,4 +546,20 @@ func TestCreateApiResolver_createArgsSCQueryService(t *testing.T) {
 		require.NotNil(t, wasmvm)
 		require.Equal(t, "*hostCore.vmHost", fmt.Sprintf("%T", wasmvm))
 	})
+	t.Run("metachain chain should NOT add systemVM", func(t *testing.T) {
```
Comment (mariusmihaic) [inline factory/api/apiResolverFactory_test.go:549]:
```suggestion
	t.Run("regular chain for meta should only add systemVM", func(t *testing.T) {
```

**Review by mariusmihaic — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**